libxmf/XmfMuxer.cpp

```diff
@@ -0,0 +1,497 @@
+	if (vid_track)
+		muxer_segment.CuesTrack(vid_track);
+	if (aud_track)
+		muxer_segment.CuesTrack(aud_track);
+
+	// Write clusters
+	unsigned char* data = NULL;
+	long data_len = 0;
+
+	const mkvparser::Cluster* cluster = parser_segment->GetFirst();
+
+	while (cluster != NULL && !cluster->EOS()) {
+		const mkvparser::BlockEntry* block_entry;
+
+		long status = cluster->GetFirst(block_entry);
+
+		if (status) {
+			printf("\n Could not get first block of cluster.\n");
+			return EXIT_FAILURE;
+		}
+
+		while (block_entry != NULL && !block_entry->EOS()) {
+			const mkvparser::Block* const block = block_entry->GetBlock();
+			const long long trackNum = block->GetTrackNumber();
+			const mkvparser::Track* const parser_track =
+				parser_tracks->GetTrackByNumber(static_cast<unsigned long>(trackNum));
+
+			// When |parser_track| is NULL, it means that the track number in the
+			// Block is invalid (i.e.) the was no TrackEntry corresponding to the
+			// track number. So we reject the file.
+			if (!parser_track) {
+				return EXIT_FAILURE;
+			}
+
+			const long long track_type = parser_track->GetType();
+			const long long time_ns = block->GetTime(cluster);
+
+			if ((track_type == Track::kAudio) ||
+				(track_type == Track::kVideo)) {
+				const int frame_count = block->GetFrameCount();
+
+				for (int i = 0; i < frame_count; ++i) {
+					const mkvparser::Block::Frame& frame = block->GetFrame(i);
+
+					if (frame.len > data_len) {
+						delete[] data;
+						data = new unsigned char[frame.len];
+						if (!data)
+							return EXIT_FAILURE;
+						data_len = frame.len;
+					}
+
+					if (frame.Read(&reader, data))
+						return EXIT_FAILURE;
+
+					mkvmuxer::Frame muxer_frame;
+					if (!muxer_frame.Init(data, frame.len))
+						return EXIT_FAILURE;
+					muxer_frame.set_track_number(track_type == Track::kAudio ? aud_track
+						: vid_track);
+					if (block->GetDiscardPadding())
+						muxer_frame.set_discard_padding(block->GetDiscardPadding());
+					muxer_frame.set_timestamp(time_ns);
+					muxer_frame.set_is_key(block->IsKey());
+					if (!muxer_segment.AddGenericFrame(&muxer_frame)) {
+						printf("\n Could not add frame.\n");
+						return EXIT_FAILURE;
+					}
+				}
+			}
+
+			status = cluster->GetNext(block_entry, block_entry);
+
+			if (status) {
+				printf("\n Could not get next block of cluster.\n");
+				break;
+			}
+		}
+
+		cluster = parser_segment->GetNext(cluster);
+	}
+
+	// We have exhausted all video and audio frames in the input file.
+	// Flush any remaining metadata frames to the output file.
+
+	const double input_duration =
+		static_cast<double>(segment_info->GetDuration()) / timeCodeScale;
+	muxer_segment.set_duration(input_duration);
+	if (!muxer_segment.Finalize()) {
+		printf("Finalization of segment failed.\n");
+		return EXIT_FAILURE;
+	}
+
+	reader.Close();
+	writer.Close();
+
+	delete[] data;
+
+	return EXIT_SUCCESS;
+}
+
+// Clean up the muxer
+void XmfWebMMuxer_Cleanup(XmfWebMMuxer* muxer) {
```

Comment:
to follow the C convention we're using for constructors and destructor mappings:

_New: allocates the object and returns a pointer
_Free: deallocates the object from its pointer (calls delete for C++)

You'd need an XmfWebMMuxer_Free function here with C# bindings so we can properly dispose of the object from managed code

Reply:
updated